File: src/lib/code/mcs-216/section-1/session-1/6/6.py
def selection_sort_detailed(arr):
    n = len(arr)
    # Metric trackers
    exchanges, comparisons = 0, 0
    outer_iters, inner_iters = 0, 0

    # Outer loop: Moves the boundary of the unsorted subarray
    for i in range(n):
        outer_iters += 1

        # Step 1: Assume the first unsorted element is the smallest
        min_index = i

        # Step 2: Inner loop to find the actual minimum in the remaining unsorted list
        for j in range(i + 1, n):
            inner_iters += 1
            comparisons += 1
            if arr[j] < arr[min_index]:
                # We found a new minimum; just update the index (no swap yet!)
                min_index = j

        # Step 3: Swap the found minimum with the first unsorted element
        # This happens ONLY once per outer loop iteration.
        if min_index != i:
            arr[i], arr[min_index] = arr[min_index], arr[i]
            exchanges += 1

    return exchanges, comparisons, outer_iters, inner_iters


# Run the analysis
data = [55, 25, 15, 40, 60, 35, 17, 65, 75, 10]
ex, comp, out, inn = selection_sort_detailed(data)
print(f"Exchanges: {ex}, Comparisons: {comp}")

File: src/lib/code/mcs-216/section-1/session-1/6/6.rs
fn selection_sort_analysis(arr: &mut [i32]) {
    let n = arr.len();
    let (mut ex, mut comp) = (0, 0);
    let (mut out, mut inn) = (0, 0);

    // Iterating through the array
    for i in 0..n {
        out += 1;
        let mut min_idx = i;

        // The inner loop scans the right side of the current index 'i'
        for j in i + 1..n {
            inn += 1;
            comp += 1;
            // Update min_idx if a smaller value is found
            if arr[j] < arr[min_idx] {
                min_idx = j;
            }
        }

        // Only swap if a new minimum was actually found at a different position
        if min_idx != i {
            arr.swap(i, min_idx);
            ex += 1;
        }
    }
    println!(
        "Exchanges: {}, Comparisons: {}, Inner Loops: {}",
        ex, comp, inn
    );
}
